### ISSUE TITLE
Fully support "Baird-style help-retractor" family

### ISSUE THREAD

**Issue by andrew-buchanan** [miguel-ambrona/deadpos#30]
Name:

    "Baird-style help-retractor".

How to identify:

    Stip begins with (usually 1 or 2) retractions, then forward stip (e.g. h#1.0 or h#1.5)

Nature:

    Formally, these are retros, and will set the "retro?" flag. 
    However there is very little retro logic required up front.
    The "help" in "help-retractor" refers to the retractions. The forward play can be co-operative or adversarial.
    There is considerable variation in forward play. Adversarial play may include s# & hs#, half-duplex & occasionally flips.
    Both retractions & forward play may include constraints (e.g. no queens, no uncaptures, checks, final move by certain piece, etc). 
    Forward aim is usually #, so DP has very limited involvement.

Performance

    Slow = "r2 >>= h#1" & greater
    Suggest best to do no initial validation of retractions (except ensure well-formed). When candidate solutions found (after applications of all constraints) can determine for each of these (relatively small number) whether any retractions or forward moves are invalid.

Desired output apart from stip:

    In the event that any solutions are identified as invalid, should post these.
    [See PDB3 examples in PDB for details.]

**Comment by andrew-buchanan** [comment]
r2 >>= h#1 should be very fast, but it's slow right now. The idea of not bothering to check retraction legality (just well-formedness) should make this as fast as h#2, and is worth implementing right now.
Recall: well-formedness = 1 king/side + no pawns on ranks 1 or 8 + no imaginary check + castling rights optional only if K&R on home square + FEN e.p.

**Comment by miguel-ambrona** [comment]
Note that there are way more retractions than forward moves, so r2 is not equivalent to m2.
Also, my implementation of retraction is not as performant as the Stockfish move generator that we use for forward play.

**Comment by andrew-buchanan** [comment]
Yes r2>h~1 is a very common stipulation, and deadpos is now fast enough on that, thank you so much.
r3+ are so much rarer, can pretty much ignore.
